File: 2023/python/31.py
with open("../test-cases/inp3.txt", "r") as inpfile:
    lines = inpfile.readlines()
    sumofpartno = 0
    symbols = ["-", "#", "=", "*", "+", "@", "$", "&", "/", "%"]
    
    def checkfornum(mL, j):
        
        try:
            if mL[j:j+3].isdigit():
                num = mL[j:j+3]
            elif mL[j:j+2].isdigit() and not mL[j-1].isdigit():
                num = mL[j:j+2]
            elif mL[j].isdigit() and not mL[j-1].isdigit():
                num = mL[j]
            else:
                return ""
        except IndexError:
            pass
        
        return num
    
    for i in range(len(lines)):
        pL = lines[i-1] if i-1 >= 0 else ""
        mL = lines[i]
        nL = lines[i+1] if i+1 < len(lines) else ""
        
        for j in range(len(mL)):
            num = checkfornum(mL, j)
            if num != "":
                if (j-1 >= 0 and (mL[j-1] in symbols)) or (j+len(num) < len(lines) and (mL[j+len(num)] in symbols)):
                    sumofpartno += int(num)
                    print(num)
                else:
                    for k in range(max(j-1, 0), min(j+len(num)+1, len(lines))):
                        if (k < len(pL) and pL[k] in symbols) or (k < len(nL) and nL[k] in symbols):
                            sumofpartno += int(num)
                            print(num)
                            break

print(f"Sum of part numbers is: {sumofpartno}")
inpfile.close()
                    
         

File: 2023/cpp/31.cpp
#include <bits/stdc++.h>
using namespace std;

string checkfornum(string mL, int j);

bool isnum(string s) {
    for (char c : s) if (!isdigit(c)) return false;
    return true;
}

bool search(char target) {
    vector<string> symbols = {
        "-", "#", "=", "*", "+", "@", "$", "&", "/", "%"};
    string targetStr(1, target);

    auto it = find(symbols.begin(), symbols.end(), targetStr);
    if (it != symbols.end()) return true;
    else return false;
}

int main(void)
{
    ifstream inpfile;
    inpfile.open("../test-cases/inp3.txt");

    int sumofpartno = 0;
    string line; 
    vector<string> lines;

    while(getline(inpfile, line)) lines.emplace_back(line);

    for (int i = 0; i < lines.size(); i++) {
        string pL = i-1 >= 0 ? lines[i-1] : "";
        string mL = lines[i];
        string nL = i+1 < lines.size() ? lines[i+1] : "";

        for (int j = 0; j < mL.size(); j++) {
            string num = checkfornum(mL, j);
            if (num != "") {
                if ((j-1>=0 && search(mL[j-1])) || (j+num.size() < mL.size() && search(mL[j+num.size()]))) { 
                    sumofpartno += stoi(num);
                    cout << num << endl;
                }
                else {
                    int limit = j+1+num.size();
                    for (int k = max(j-1, 0); k < min(limit, static_cast<int>(line.size())); k++) {
                        if (k<pL.size() && search(pL[k])) {
                            sumofpartno += stoi(num);
                            cout << num << endl;
                            break;
                        }
                        else if (k<nL.size() && search(nL[k])) {
                            sumofpartno += stoi(num);
                            cout << num << endl;
                            break;
                        }
                    }
                }
            }
        }
    }

    cout << "Sum of part numbers: " << sumofpartno << endl;
    inpfile.close();
}

string checkfornum(string mL, int j) {
    string num = "";
    try {
        if (isnum(mL.substr(j,3))) num = mL.substr(j,3);
        else if (isnum(mL.substr(j,2)) && !isdigit(mL[j-1])) num = mL.substr(j,2);
        else if (isdigit(mL[j]) && !isdigit(mL[j-1])) num = mL[j];
    } catch (const out_of_range&) {}

    return num;
}

